akka-http-core/src/main/scala/akka/http/impl/engine/ws/WebSocket.scala

```diff
@@ -51,12 +56,35 @@ private[http] object WebSocket {
     Masking(serverSide, maskingRandomFactory)
       .named("ws-masking")
 
+  /** The layer that transparently injects (if enabled) keepAlive Ping or Pong messages when connection is idle */
+  def periodicKeepAlive(settings: WebSocketSettings): BidiFlow[FrameHandler.Output, FrameHandler.Output, FrameOutHandler.Input, FrameOutHandler.Input, NotUsed] = {
+    settings.periodicKeepAliveMaxIdle match {
+      case maxIdle: FiniteDuration ⇒
+
+        val noCustomData = WebSocketSettingsImpl.hasNoCustomPeriodicKeepAliveData(settings)
+        val mkFrame = settings.periodicKeepAliveMode match {
+          case "ping" if noCustomData ⇒ FrameHandler.mkDirectAnswerPing // sending Ping should result in a Pong back
+          case "ping" ⇒ () ⇒ DirectAnswer(FrameEvent.fullFrame(Opcode.Ping, None, settings.periodicKeepAliveData(), fin = true))
+
+          case "pong" if noCustomData ⇒ FrameHandler.mkDirectAnswerPong // sending Pong means we do not expect a reply
+          case "pong"                 ⇒ () ⇒ DirectAnswer(FrameEvent.fullFrame(Opcode.Pong, None, settings.periodicKeepAliveData(), fin = true))
```

Comment:
better error message if not matching ping or pong?

Reply:
ok!